pumpkin-inventory/src/lib.rs

```diff
@@ -206,32 +215,51 @@ impl<'a> Container for OptionallyCombinedContainer<'a, 'a> {
     }
 
     fn window_name(&self) -> &'static str {
-        self.container
-            .as_ref()
-            .map(|container| container.window_name())
-            .unwrap_or(self.inventory.window_name())
-    }
-
-    fn all_slots(&mut self) -> Vec<&mut Option<ItemStack>> {
-        let slots = match &mut self.container {
-            Some(container) => {
-                let mut slots = container.all_slots();
-                slots.extend(self.inventory.all_combinable_slots_mut());
-                slots
-            }
-            None => self.inventory.all_slots(),
+        let Some(container) = &self.container else {
+            return self.inventory.window_name();
         };
-        slots
+        container.window_name()
     }
 
-    fn all_slots_ref(&self) -> Vec<Option<&ItemStack>> {
+    fn iter_slots_mut<'s>(&'s mut self) -> Box<(dyn Iterator<Item = &mut Option<ItemStack>> + 's)> {
+        match &mut self.container {
+            Some(container) => Box::new(
+                container
+                    .iter_slots_mut()
+                    .chain(self.inventory.iter_slots_mut()),
+            ),
+            None => self.inventory.iter_slots_mut(),
+        }
+    }
+
+    fn iter_slots<'s>(&'s self) -> Box<(dyn Iterator<Item = &Option<ItemStack>> + 's)> {
         match &self.container {
-            Some(container) => {
-                let mut slots = container.all_slots_ref();
-                slots.extend(self.inventory.all_combinable_slots());
-                slots
-            }
-            None => self.inventory.all_slots_ref(),
+            Some(container) => Box::new(container.iter_slots().chain(self.inventory.iter_slots())),
+            None => self.inventory.iter_slots(),
         }
     }
+
+    fn get_slot(&self, slot: usize) -> Option<&Option<ItemStack>> {
+        if (0..self.inventory.size()).contains(&slot) {
+            return self.inventory.get_slot(slot);
+        }
+        if let Some(container) = &self.container {
+            return container.get_slot(slot);
```

Comment:
you need to `return container.get_slot(slot-self.inventory.size())`

Reply:
actually nevermind, this should be a match statement, looking something like:
```rust
match &self.container {
    Some(container) => {
        if (0..container.size()).contains(&slot) {
            container.get_slot(slot)
        } else {
            self.inventory.items.get(slot)
        }
    None => self.inventory.get_slot(slot)
}
```

see https://wiki.vg/Inventory#Chest